ic-agent/src/agent/replica_api.rs

```diff
@@ -146,8 +147,53 @@ pub enum QueryResponse {
     #[serde(rename = "replied")]
     Replied { reply: CallReply },
     #[serde(rename = "rejected")]
-    Rejected {
-        reject_code: u64,
-        reject_message: String,
-    },
+    Rejected(RejectResponse),
+}
+
+/// An IC execution error received from the replica.
+#[derive(Debug, Clone, Serialize, Deserialize, Ord, PartialOrd, Eq, PartialEq)]
+pub struct RejectResponse {
+    /// The [reject code](https://smartcontracts.org/docs/interface-spec/index.html#reject-codes) returned by the replica.
+    pub reject_code: RejectCode,
+    /// The rejection message.
+    pub reject_message: String,
+    /// The optional [error code](https://smartcontracts.org/docs/interface-spec/index.html#error-codes) returned by the replica.
+    #[serde(default)]
+    pub error_code: Option<String>,
+}
+
+/// See the [interface spec](https://smartcontracts.org/docs/interface-spec/index.html#reject-codes).
+#[derive(
+    Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize_repr, Deserialize_repr, Ord, PartialOrd,
+)]
+#[repr(u8)]
+pub enum RejectCode {
+    /// Fatal system error, retry unlikely to be useful
+    SysFatal = 1,
+    /// Transient system error, retry might be possible.
+    SysTransient = 2,
+    /// Invalid destination (e.g. canister/account does not exist)
+    DestinationInvalid = 3,
+    /// Explicit reject by the canister.
+    CanisterReject = 4,
+    /// Canister error (e.g., trap, no response)
+    CanisterError = 5,
+}
+
+impl TryFrom<u64> for RejectCode {
+    type Error = AgentError;
+
+    fn try_from(value: u64) -> Result<Self, AgentError> {
+        match value {
+            1 => Ok(RejectCode::SysFatal),
+            2 => Ok(RejectCode::SysTransient),
+            3 => Ok(RejectCode::DestinationInvalid),
+            4 => Ok(RejectCode::CanisterReject),
+            5 => Ok(RejectCode::CanisterError),
+            _ => Err(AgentError::MessageError(format!(
```

Comment:
I don't think the MessageError type is very appropriate here.

1.  The description reads -  #[error("A tool returned a string message error: {0}")]. There is actually no tool, but also it seems we never set this error in other place as well.
2. I think CallDataMismatch is more appropriate error.

If you agree I would suggest fixing this and at the same time even remove the AgentError::MessageError field. The error space is already too big for users of the IC agent to handle it gracefully. So no point in including error types that are not set.